optax/assignment/_hungarian_algorithm.py

```diff
@@ -0,0 +1,350 @@
+
+  Args:
+    costs: A matrix of costs.
+
+  Returns:
+    A pair ``(i, j)`` where ``i`` is an array of row indices and ``j`` is an
+    array of column indices.
+    The cost of the assignment is ``costs[i, j].sum()``.
+
+  References:
+    David F. Crouse. [On implementing 2D rectangular assignment algorithms](
+     https://ieeexplore.ieee.org/document/7738348). IEEE Transactions on
+     Aerospace and Electronic Systems. 52(4):1679-1696, August 2016.
+    https://en.wikipedia.org/wiki/Linear_assignment_problem
+    https://en.wikipedia.org/wiki/Hungarian_algorithm
+
+  Examples:
+    >>> import optax
+    >>> from jax import numpy as jnp
+    >>> cost = jnp.array(
+    ...  [
+    ...    [8, 4, 7],
+    ...    [5, 2, 3],
+    ...    [9, 6, 7],
+    ...    [9, 4, 8],
+    ...  ])
+    >>> i, j = optax.assignment.hungarian_algorithm(cost)
+    >>> print("cost:", cost[i, j].sum())
+    cost: 15
+    >>> print("i:", i)
+    i: [0 1 3]
+    >>> print("j:", j)
+    j: [0 2 1]
+    >>> cost = jnp.array(
+    ...  [
+    ...    [90, 80, 75, 70],
+    ...    [35, 85, 55, 65],
+    ...    [125, 95, 90, 95],
+    ...    [45, 110, 95, 115],
+    ...    [50, 100, 90, 100],
+    ...  ])
+    >>> i, j = optax.assignment.hungarian_algorithm(cost)
+    >>> print("cost:", cost[i, j].sum())
+    cost: 265
+    >>> print("i:", i)
+    i: [0 1 2 3]
+    >>> print("j:", j)
+    j: [3 2 1 0]
+  """
+
+  if costs.shape[0] == 0 or costs.shape[1] == 0:
+    return jnp.zeros(0, int), jnp.zeros(0, int)
+
+  transpose = costs.shape[1] < costs.shape[0]
+
+  if transpose:
+    costs = costs.T
+
+  costs = costs.astype(float)
+  u = jnp.zeros(costs.shape[0], costs.dtype)
+  v = jnp.zeros(costs.shape[1], costs.dtype)
+
+  path = jnp.full(costs.shape[1], -1)
+  col4row = jnp.full(costs.shape[0], -1)
+  row4col = jnp.full(costs.shape[1], -1)
+
+  init = costs, u, v, path, row4col, col4row
+  costs, u, v, path, row4col, col4row = lax.fori_loop(
+    0, costs.shape[0], _lsa_body, init
+  )
+
+  if transpose:
+    i = col4row.argsort()
+    return col4row[i], i
+  else:
+    return jnp.arange(costs.shape[0]), col4row
+
+
+def _find_short_augpath_while_body_inner_for(it, val):
+  (
+    remaining,
+    min_value,
+    costs,
+    i,
+    u,
+    v,
+    shortest_path_costs,
+    path,
+    lowest,
+    row4col,
+    index,
+  ) = val
+
+  j = remaining[it]
+
+  r = min_value + costs[i, j] - u[i] - v[j]
+
+  path = path.at[j].set(jnp.where(r < shortest_path_costs[j], i, path[j]))
+
+  shortest_path_costs = shortest_path_costs.at[j].min(r)
+
+  index = jnp.where(
+    (shortest_path_costs[j] < lowest)
+    | ((shortest_path_costs[j] == lowest) & (row4col[j] == -1)),
+    it,
+    index,
+  )
+
+  lowest = jnp.minimum(lowest, shortest_path_costs[j])
+
+  return (
+    remaining,
+    min_value,
+    costs,
+    i,
+    u,
+    v,
+    shortest_path_costs,
+    path,
+    lowest,
+    row4col,
+    index,
+  )
+
+
+def _find_short_augpath_while_body_tail_alt(val):
```

Comment:
All names of functions are very well chosen, which makes the code nicely readable. An exception is "alt" here: is it for alternative? Either change this word to something readable or add a comment below

Reply:
Ah, that's supposed to be just `_find_short_augpath_while_body_tail`. (I was testing an alternative implementation.)